[Winson Heng] iP

winsonheng opened the pull request:
### Duke

> "CS2103T is a low workload mod" – Albert Einstein

- [ ] Lightweight
- [x] Text-based _(great for programmers!)_
- [ ] Easy to use (just run `Duke.main()`)

**Upcoming features:**

- JavaFX GUI
- ~~JavaFX will definitely be useful for my career~~

Honestly you should just use [any other todo app](https://zapier.com/blog/best-todo-list-apps/):

1. Todoist
2. Microsoft To Do
#
Crazy how the entire program is just 3 lines of code!  
```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

winsonheng committed:
Completed Level 1

winsonheng committed:
Updated gitignore to ignore class files

winsonheng committed:
Revert "Updated gitignore to ignore class files"

This reverts commit f8a5dc4184db1070e011d3a5a3ae2e97422faec2.

winsonheng committed:
Revert "Completed Level 1"

This reverts commit 80b65a70f7b8fdc73e74040b2d303bd8ee0c5a7f.

winsonheng committed:
Completed Level 1

winsonheng committed:
Completed Level 2

winsonheng committed:
Completed Level-3

winsonheng committed:
Completed Level 4

winsonheng committed:
Completed A-TextUiTesting

winsonheng committed:
Completed Level 5

winsonheng committed:
Completed Level 6

winsonheng committed:
Completed A-Enums

XylusChen commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,98 @@
 public class Duke {
+    private static final String FULL_LINE = "_______________________________________________\n";
+    private static final String ADD_TASK_OUTPUT = "Got it. I've added this task:\n\t%s\nNow you have %d tasks in the list.";
+    private static final String DELETE_TASK_OUTPUT = "Noted. I've removed this task:\n\t%s\nNow you have %d tasks in the list.";
```
I like the usage of constant variables here. Makes referencing these commonly used outputs easy! 👍

XylusChen commented on src/main/java/ERROR.java:
```diff
@@ -0,0 +1,19 @@
+package main.java;
+
+public enum ERROR {
```
Great use of `enum` here to make instantializing DukeExceptions easier! 👍

XylusChen commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,98 @@
+                    Task task = taskList.get(index - 1);
+                    task.markAsUndone();
+                    output = "OK, I've marked this task as not done yet:\n" + task.toString();
+                } else if (input.matches("delete \\d+")) {
+                    int index = Integer.parseInt(input.replace("delete ", ""));
+                    if (index == 0 || index > taskList.size()) {
+                        throw new DukeException(String.format(ERROR.INVALID_INDEX.getMessage(), taskList.size()));
+                    }
+                    Task task = taskList.get(index - 1);
+                    taskList.remove(index - 1);
+                    output = String.format(DELETE_TASK_OUTPUT, task.toString(), taskList.size());
+                } else if (input.startsWith("todo")) {
+                    String desc = input.replace("todo", "").trim();
+                    if (desc.isBlank()) {
+                        throw new DukeException(ERROR.TODO_EMPTY.getMessage());
+                    }
+                    Todo task = new Todo(desc);
+                    taskList.add(task);
+                    output = String.format(ADD_TASK_OUTPUT, task.toString(), taskList.size());
+                } else if (input.startsWith("deadline")) {
+                    String desc = input.replace("deadline", "").trim();
+                    if (desc.isBlank()) {
+                        throw new DukeException(ERROR.DEADLINE_EMPTY.getMessage());
+                    }
+                    String[] params = desc.split(" /by ");
+                    Deadline task = new Deadline(params[0], params[1]);
+                    taskList.add(task);
+                    output = String.format(ADD_TASK_OUTPUT, task.toString(), taskList.size());
+                } else if (input.startsWith("event")) {
+                    String desc = input.replace("event", "").trim();
+                    if (desc.isBlank()) {
+                        throw new DukeException(ERROR.EVENT_EMPTY.getMessage());
+                    }
+                    String[] params = desc.split("( /from | /to )");
+                    Event task = new Event(params[0], params[1], params[2]);
+                    taskList.add(task);
+                    output = String.format(ADD_TASK_OUTPUT, task.toString(), taskList.size());
+                } else {
+                    throw new DukeException(ERROR.INVALID_INPUT.getMessage());
+                }
```
Perhaps the logic here could be abstracted out❓

XylusChen replied:
I like how you used inbuilt String methods to parse input commands. 💯

XylusChen reviewed (commented):
LGTM. Overall very-well written code. Easily readable and comprehendable. Good job!

winsonheng committed:
Completed Level 7

winsonheng committed:
Completed Level 8

winsonheng committed:
Merged Level 7 and 8

winsonheng committed:
Delete Deadline.java.orig

winsonheng committed:
Completed More OOP

winsonheng committed:
Completed A-Packages

winsonheng committed:
Merge remote-tracking branch 'origin/add-gradle-support'

winsonheng committed:
Completed A-Gradle

winsonheng committed:
Completed A-JUnit

winsonheng committed:
Completed A-Jar

winsonheng committed:
Completed A-JavaDoc

winsonheng committed:
Completed A-CodingStandard

winsonheng committed:
Completed Level-9

winsonheng committed:
Merged A-JavaDoc and A-CodingStandard

winsonheng committed:
Merged with Level-9

winsonheng committed:
Completed Level-10

winsonheng committed:
Generate new JAR file with JavaFX GUI

duke.jar is a command line app from Week 3.

The new duke.jar file allows the JavaFX GUI app to be used directly
without compiling the code.

winsonheng committed:
Add assert statements in task classes

These new assert statements ensure the current state is valid,
especially when parsing, marking or deleting tasks, to prevent cases
where the task is null or the index is invalid.

winsonheng committed:
Add assert statements to Command

Forgot to branch previously so assert statements are already added to
task classes.

winsonheng committed:
Refactor code to improve code quality

winsonheng committed:
Merge pull request #1 from winsonheng/branch-A-CodeQuality

Refactor code to improve code quality

winsonheng committed:
Add classes for text and csv storage formats

winsonheng committed:
Add menu item to select and load csv file

winsonheng committed:
Merge branch 'master' into branch-C-FlexibleDataSource

winsonheng committed:
Merge pull request #2 from winsonheng/branch-C-FlexibleDataSource

Branch c flexible data source

winsonheng committed:
Improve GUI

winsonheng committed:
Final Changes to GUI and DateTime formatting

winsonheng committed:
Update README.md

winsonheng committed:
Merge pull request #4 from winsonheng/branch-A-BetterGui

Improve GUI

winsonheng committed:
Merge pull request #5 from winsonheng/branch-A-Personality

Final Changes to GUI and DateTime formatting

winsonheng committed:
Create release v0.2

winsonheng committed:
Update README.md

winsonheng committed:
Update README.md

winsonheng committed:
Edit bye command to quit app in 3 seconds

winsonheng committed:
Add tests for Task

winsonheng committed:
Update README.md
